ginga/web/pgw/Widgets.py

```diff
@@ -1918,6 +2465,66 @@ def add_widget(self, child, row, col, stretch=0):
                              value=self.render_body())
         self.make_callback('widget-added', child)
 
+    def insert_cell(self, row, col):
+        indices = [row, col]
+        if self._rendered:
+            app = self.get_app()
+            app.do_operation("insert_cell", id=self.id, value=indices)
+
+    def insert_row(self, index):
+        indices = [index, self.num_cols]
+        self.num_rows += 1
+
+        # handle case where user inserts row at the end of the gridbox
+        if index == self.num_rows - 1:
+            for j in range(self.num_cols):
+                self.tbl[(index, j)] = Box()
+        else:
+            # shift key/value pairs down to make the row empty at index
+            for i in range(self.num_rows - 2, index - 1, -1):
+                for j in range(self.num_cols):
+                    self.tbl[(i + 1, j)] = self.tbl[(i, j)]
+            # populate inserted row with empty Boxes for render_body()
+            for j in range(self.num_cols):
+                self.tbl[(index, j)] = Box()
+
+        if self._rendered:
+            app = self.get_app()
+            app.do_operation("insert_row", id=self.id, value=indices)
+
+    def append_row(self):
+        indices = [self.num_rows, self.num_cols]
+        if self._rendered:
+            app = self.get_app()
+            app.do_operation("append_row", id=self.id, value=indices)
+        self.num_rows += 1
+        # populate appended row with empty Boxes for render_body()
+        for j in range(self.num_cols):
+            self.tbl[(self.num_rows - 1, j)] = Box()
+
+    def delete_row(self, index):
+
+        if index < 0 or index >= self.num_rows:
+            print("Index out of bounds")
```

Comment:
Leftover debug statement? If not, why not raise exception here or use the logger?

```suggestion
```